src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -20,110 +22,157 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine
     public class ProxyTestSessionManager : IProxyTestSessionManager
     {
         private readonly object lockObject = new object();
-        private int parallelLevel;
-        private bool skipDefaultAdapters;
+        private readonly object proxyOperationLockObject = new object();
+        private StartTestSessionCriteria testSessionCriteria;
+        private int testhostCount;
+        private TestSessionInfo testSessionInfo;
         private Func<ProxyOperationManager> proxyCreator;
-        private Queue<Guid> availableProxyQueue;
-        private IDictionary<Guid, ProxyOperationManagerContainer> proxyMap;
+        private IList<ProxyOperationManagerContainer> proxyContainerList;
+        private IDictionary<string, int> proxyMap;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ProxyTestSessionManager"/> class.
         /// </summary>
         /// 
-        /// <param name="parallelLevel">The parallel level.</param>
+        /// <param name="criteria">The test session criteria.</param>
+        /// <param name="testhostCount">The testhost count.</param>
         /// <param name="proxyCreator">The proxy creator.</param>
-        public ProxyTestSessionManager(int parallelLevel, Func<ProxyOperationManager> proxyCreator)
+        public ProxyTestSessionManager(
+            StartTestSessionCriteria criteria,
+            int testhostCount,
+            Func<ProxyOperationManager> proxyCreator)
         {
-            this.parallelLevel = parallelLevel;
+            this.testSessionCriteria = criteria;
+            this.testhostCount = testhostCount;
             this.proxyCreator = proxyCreator;
 
-            this.availableProxyQueue = new Queue<Guid>();
-            this.proxyMap = new Dictionary<Guid, ProxyOperationManagerContainer>();
+            this.proxyContainerList = new List<ProxyOperationManagerContainer>();
+            this.proxyMap = new Dictionary<string, int>();
         }
 
         /// <inheritdoc/>
-        public void Initialize(bool skipDefaultAdapters)
-        {
-            this.skipDefaultAdapters = skipDefaultAdapters;
-        }
-
-        /// <inheritdoc/>
-        public void StartSession(
-            StartTestSessionCriteria criteria,
-            ITestSessionEventsHandler eventsHandler)
+        public void StartSession(ITestSessionEventsHandler eventsHandler)
         {
-            var testSessionInfo = new TestSessionInfo();
-            Task[] taskList = new Task[this.parallelLevel];
+            lock (this.lockObject)
+            {
+                if (this.testSessionInfo != null)
+                {
+                    return;
+                }
+                this.testSessionInfo = new TestSessionInfo();
+            }
 
             // Create all the proxies in parallel, one task per proxy.
-            for (int i = 0; i < this.parallelLevel; ++i)
+            var taskList = new Task[this.testhostCount];
+            for (int i = 0; i < this.testhostCount; ++i)
```

Comment:
What if in big test solutions we will have  100+ test projects? How performance behaves in that case?
Is it reasonable to run all projects in parallel? Is it better to have some limitation?